### PR TITLE
Lead CRUD For Tenant

### PR DESCRIPTION
**Pull request by komal-wes**
Create CRUD to manage Leads for Tenant

### CODE REVIEW

File: app/Constants/Lead/StatusConstants.php [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        return [
+            self::NEW,
+            self::OPEN,
+            self::CLOSED
+        ];
+    }
+
+    public static function getStatusType($value)
+    {
+        switch ($value) {
```
Comment (SatwinderWebethics) [inline app/Constants/Lead/StatusConstants.php]: Please don't use switch statement, There is no need to use the switch statement

File: resources/views/app/leads/index.blade.php [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+                                <tr>
+                                    <td class="px-6 py-4 whitespace-nowrap">
+                                        <div class="text-sm text-gray-900">{{ $lead->lead_date }}</div>
+                                    </td>
+                                    <td class="px-6 py-4 whitespace-nowrap">
+                                        <div class="text-sm text-gray-900">{{ $lead->client_name }}</div>
+                                    </td>
+                                    <td class="px-6 py-4 whitespace-nowrap">
+                                        <div class="text-sm text-gray-900">
+                                            {{ \App\Constants\Lead\SourceConstants::getSourceType($lead->lead_source) }}
```
Comment (SatwinderWebethics) [inline resources/views/app/leads/index.blade.php]: Send it from the controoler

File: routes/tenant.php [hunk 3]
```diff
@@ -30,5 +27,9 @@
 ])->group(function () {
     Route::get('/', [DashboardController::class, 'index'])->name('tenant.home');
     Route::get('/tenant-dashboard', [DashboardController::class, 'dashboard'])->middleware(['auth', 'verified'])->name('tenant.dashboard');
+
+    Route::post('leads/update-lead/{lead}', [LeadController::class, 'updateLead'])->name('leads.updateLead');
```
Comment (SatwinderWebethics) [inline routes/tenant.php:31]: What is the use of this route?

File: database/migrations/tenant/2024_04_05_064217_create_leads_table.php [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+     */
+    public function up(): void
+    {
+        Schema::create('leads', function (Blueprint $table) {
+            $table->id();
+            $table->unsignedBigInteger('user_id');
+            $table->foreign('user_id')->references('id')->on('users');
+            $table->string('client_name');
+            $table->integer('lead_source');
+            $table->timestamp('lead_date')->nullable();
```
Comment (SatwinderWebethics) [inline database/migrations/tenant/2024_04_05_064217_create_leads_table.php]: Take it as date type column

File: app/Http/Controllers/Tenant/LeadController.php [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        $statuses = StatusConstants::getStatuses();
+        return view('app.leads.create', ['lead_sources' => $lead_sources,'statuses' => $statuses,'lead' => $lead]);
+    }
+
+    /**
+     * Remove the specified resource from storage.
+     */
+    public function destroy(string $id)
+    {
+        $lead = Lead::find($id);
```
Comment (SatwinderWebethics) [inline app/Http/Controllers/Tenant/LeadController.php]: Remove find function there is no need to find and delete, search for better way to how can be be achieve the same thing without find function, and also do same for update function

File: routes/tenant.php (hunk 3 above)
Comment (komal-wes) [reply routes/tenant.php:31]: This route is being used to update the lead on edit.

File: app/Http/Controllers/Tenant/LeadController.php [hunk 6]
```diff
@@ -69,10 +66,7 @@ public function show(Lead $lead)
      */
     public function destroy(string $id)
     {
-        $lead = Lead::find($id);
-        if ($lead) {
-            $lead->delete();
-        }
+        $deleted = Lead::destroy($id);
```
Comment (SatwinderWebethics) [inline app/Http/Controllers/Tenant/LeadController.php:61]: It work if id not found?

File: app/Models/Lead.php [hunk 7]
```diff
@@ -39,4 +43,14 @@ protected function leadDate(): Attribute
             get: fn (string $value) => Carbon::parse($value)->toDateString(),
         );
     }
+
+    public function getLeadSourceTypeAttribute()
+    {
```
Comment (SatwinderWebethics) [inline app/Models/Lead.php:56]: Can you try it without appending the things?

File: resources/views/app/leads/create.blade.php [hunk 8]
```diff
@@ -81,7 +81,7 @@
                         </div>
                         <div class="flex items-center justify-end mt-4">
                             <x-primary-button class="ms-4">
-                                {{ __('Add Lead') }}
+                                {{ __(isset($lead)?'Update':'Save') }}
```
Comment (SatwinderWebethics) [inline resources/views/app/leads/create.blade.php]: Just write save

File: app/Constants/Lead/SourceConstants.php [hunk 9]
```diff
@@ -9,6 +9,14 @@ class SourceConstants
     const Source3 = 2;
     const Source4 = 3;
 
+
+    private static $sourceLabels = [
+        self::Source1 => 'Source1',
+        self::Source2 => 'Source2',
+        self::Source3 => 'Source3',
+        self::Source4 => 'Source4',
```
Comment (SatwinderWebethics) [inline app/Constants/Lead/SourceConstants.php]: Please use Enum instead constants

File: app/Http/Controllers/Tenant/LeadController.php (hunk 6 above)
Comment (komal-wes) [reply app/Http/Controllers/Tenant/LeadController.php:61]: Yes!

File: resources/views/app/leads/create.blade.php (hunk 8 above)
Comment (komal-wes) [reply resources/views/app/leads/create.blade.php]: Okay

File: app/Constants/Lead/SourceConstants.php (hunk 9 above)
Comment (komal-wes) [reply app/Constants/Lead/SourceConstants.php]: Okay!

File: app/Constants/Lead/SourceConstants.php
Comment (SatwinderWebethics) [inline app/Constants/Lead/SourceConstants.php:1]: Please use propers naming, not it's not a constant it's a enum

File: app/Models/Lead.php [hunk 10]
```diff
@@ -44,13 +41,21 @@ protected function leadDate(): Attribute
         );
     }
 
-    public function getLeadSourceTypeAttribute()
-    {
-        return SourceConstants::getSourceType($this->attributes['lead_source']);
+    public function getLeadSourcesAttribute()
```
Comment (SatwinderWebethics) [inline app/Models/Lead.php]: What you are using the loops in append attribute, Please try to without appends

File: app/Enums/StatusesEnum.php [hunk 11]
```diff
@@ -1,7 +1,6 @@
 <?php
-namespace App\Constants\Lead;
-
-enum StatusConstants: int {
+namespace App\Enums;
+enum StatusesEnum: int {
     case NEW = 0;
```
Comment (SatwinderWebethics) [inline app/Enums/StatusesEnum.php]: Define the constants in models and use them.

**Pull request merged**